util/src/main/java/io/kubernetes/client/util/leaderelection/LeaderElector.java

```diff
@@ -0,0 +1,210 @@
+package io.kubernetes.client.util.leaderelection;
+
+import io.kubernetes.client.ApiException;
+import java.time.Duration;
+import java.util.Date;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LeaderElector {
+
+  private static final Logger log = LoggerFactory.getLogger(LeaderElector.class);
+
+  private static final int STATUS_NOT_FOUND = 404;
+
+  private final LeaderElectionConfig config;
+
+  // internal bookkeeping
+  private LeaderElectionRecord observedRecord;
+  private long observedTimeMilliSeconds;
+
+  private ScheduledExecutorService scheduledWorkers = Executors.newSingleThreadScheduledExecutor();
+  private ExecutorService leaseWorkers = Executors.newSingleThreadExecutor();
+  private ExecutorService hookWorkers = Executors.newSingleThreadExecutor();
+
+  public LeaderElector(LeaderElectionConfig config) {
+    this.config = config;
+  }
+
+  public void run(Runnable startLeadingHook, Runnable stopLeadingHook) {
+    log.info("Start leader election with lock", config.getLock().describe());
+    try {
+      if (!acquire()) {
+        // Fail to acquire leadership
+        return;
+      }
+      log.info("Successfully acquired lease, became leader");
+      // Hook on start leading
+      hookWorkers.submit(startLeadingHook);
+      renew();
```

Comment:
It feels weird to me that this function loops forever... I'd rather see the `while` loop here for clarity.

Reply:
how about adding a comment here: 

> // renew() won't return until it fails acquiring/renewing the lease

Reply:
I changed the method name to `renewLoop`. Is it better?